src/Archives/HuffLZ.h

```diff
@@ -31,9 +30,9 @@ namespace Archives
 		void WriteCharToBuffer(char c);
 
 		// Member variables
-		BitStreamReader *m_BitStreamReader;
-		BitStreamReader *m_ConstructedBitStreamReader;
-		AdaptiveHuffmanTree *m_AdaptiveHuffmanTree;
+		BitStreamReader m_BitStreamReader;
+		BitStreamReader m_ConstructedBitStreamReader;
```

Comment:
Ahh, I see you've changed the type of this field as well. Hmm, this field actually makes much less sense now. We should probably eliminate it. Previously the combination of `m_BitStreamReader` and `m_ConstructedBitStreamReader` together functioned somewhat like a `shared_ptr`. It would either reference an existing object which it did not own (and hence would not free), or it would allocated the object itself and thus be responsible for the lifetime and cleanup of that object.

By using containment, we are implying ownership. There is no question about it. If we are passed an object, it must be copied before it can be used. That is, for the constructor that takes the `BitStreamReader` object, the object must be copied rather than simply storing a reference to it.

In hindsight, I probably wrote the code as it was thinking there was a bigger performance gain by avoiding the copy. I had probably erroneously assumed the entire data buffer would be copied. In actuality though, it's a shallow copy, in that only the 4 fields defined in `BitStreamReader` are copied, such as the buffer pointer. It doesn't need to copy the actual data buffer itself. The copy operation is actually fairly cheap, and not frequently occurring. Performance implications should be fairly minimal. The copy operation is slightly more expensive this way, though there is less indirection when accessing the object, so some slight efficiency gains there. This might even be a net gain in terms of performance, and a big gain in terms of code readability.

My advice is to drop the `m_ConstructedBitStreamReader` field now that the referenced objects are contained in the host class.

As a side note, it would be reasonable to eliminate the alternate constructor that takes a pointer and size, and instead only take a constructed `BitStreamReader`. Such an object is cheap to construct and copy. The parameter could be `const`, since a (non-const) copy is made. That could allow a temporarily constructed object to be passed to the constructor. Eliminating the alternate constructor that takes individual pointer and size values cuts down on code, and cuts down on the number of ways the library could be used to accomplish the same task. That might actually be a usability gain.

Reply:
Side side note: The Stream code uses the parameter order `(void* buffer, std::size_t size)`, but the HuffLZ object reverses it to `(std::size_t size, void* buffer)`. We should probably try to be consistent about the ordering.